crates/provider-wadm/src/lib.rs

```diff
@@ -0,0 +1,624 @@
+// NOTE: This comes from wrpc?
+#![recursion_limit = "256"]
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::wasmcloud::wadm::wadm_types::{
+    DeleteModelResponse, DeployResponse, GetModelResponse, ModelSummary, PutModelResponse,
+    StatusResponse, VersionResponse,
+};
+use anyhow::{anyhow, bail, Context as _};
+use async_nats::subject::ToSubject;
+use async_nats::HeaderMap;
+use futures::StreamExt;
+use opentelemetry_nats::attach_span_context;
+use tokio::fs;
+use tokio::sync::{OwnedSemaphorePermit, RwLock, Semaphore};
+use tokio::task::JoinHandle;
+use tracing::{debug, error, instrument, warn};
+use tracing_futures::Instrument;
+use wascap::prelude::KeyPair;
+use wasmcloud::messaging::types::BrokerMessage;
+use wasmcloud::wadm::oam_types::OamManifest;
+use wasmcloud_provider_sdk::core::HostData;
+use wasmcloud_provider_sdk::wasmcloud_tracing::context::TraceContextInjector;
+use wasmcloud_provider_sdk::{
+    get_connection, load_host_data, run_provider, Context, LinkConfig, Provider,
+};
+
+mod config;
+use config::{WadmConfig, WADM_STATUS_API_PREFIX};
+
+mod translate;
+
+wit_bindgen_wrpc::generate!({
+    additional_derives: [
+        serde::Serialize,
+        serde::Deserialize,
+    ],
+});
+
+pub async fn run() -> anyhow::Result<()> {
+    WadmProvider::run().await
+}
+
+/// [`NatsClientBundle`]s hold a NATS client and information (subscriptions)
+/// related to it.
+///
+/// This struct is necssary because subscriptions are *not* automatically removed on client drop,
+/// meaning that we must keep track of all subscriptions to close once the client is done
+#[derive(Debug)]
+struct NatsClientBundle {
+    pub client: async_nats::Client,
+    pub sub_handles: Vec<(String, JoinHandle<()>)>,
```

Comment:
Nit: You can use the `abort_handle` method on a `JoinHandle` to get something you can abort instead. Might be slightly cleaner as you don't need to await this handle